File: framework-web/src/main/java/com/example/framework/web/interceptor/TenantRequestInterceptor.java
package com.example.framework.web.interceptor;

import com.example.framework.web.constants.TenantConstants;
import com.example.framework.web.contexts.TenantContext;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.jetbrains.annotations.NotNull;
import org.springframework.web.servlet.AsyncHandlerInterceptor;

import java.util.Optional;

/**
 * @author MaGuangZu
 * @since 2024-05-19
 */
public class TenantRequestInterceptor implements AsyncHandlerInterceptor {

	@Override
	public boolean preHandle(@NotNull HttpServletRequest request, @NotNull HttpServletResponse response, @NotNull Object handler) {
		return Optional.of(request)
			// .map(req -> securityDomain.getTenantIdFromJwt(req))
			.map(it -> {
				return setTenantContext(TenantConstants.ROOT_TENANT);
			})
			.orElse(false);
	}

	@Override
	public void afterCompletion(@NotNull HttpServletRequest request, @NotNull HttpServletResponse response, @NotNull Object handler, Exception ex) throws Exception {
		TenantContext.clear();
	}

	private Boolean setTenantContext(String tenant) {
		TenantContext.setTenantId(tenant);
		return true;
	}

}

File: framework-web/src/main/kotlin/com/example/framework/web/interceptor/TenantRequestInterceptor.kt
package com.example.framework.web.interceptor

import com.example.framework.web.constants.TenantConstants
import com.example.framework.web.contexts.TenantContext
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.lang.Nullable
import org.springframework.web.servlet.AsyncHandlerInterceptor
import java.util.*

class TenantRequestInterceptor : AsyncHandlerInterceptor {

	override fun preHandle(request: HttpServletRequest, response: HttpServletResponse, handler: Any): Boolean {
		return Optional.ofNullable(request)
			// .map(req -> securityDomain.getTenantIdFromJwt(req))
			.map { _: HttpServletRequest? -> setTenantContext(TenantConstants.ROOT_TENANT) }
			.orElse(false)
	}

	@Throws(Exception::class)
	override fun afterCompletion(
		request: HttpServletRequest,
		response: HttpServletResponse,
		handler: Any,
		@Nullable ex: java.lang.Exception?
	) {
		TenantContext.clear()
	}

	private fun setTenantContext(tenant: String): Boolean {
		TenantContext.setTenantId(tenant)
		return true
	}

}
